src/MessageBodyRenderer.php

```diff
@@ -0,0 +1,187 @@
+     *
+     * @param string $view The view name of the view file.
+     * @param array $viewParameters The parameters (name-value pairs) that will be extracted and available in
+     * the view file.
+     * @param array $layoutParameters The parameters (name-value pairs) that will be extracted and available in
+     * the layout file.
+     *
+     * @see View::render()
+     *
+     * @return string The rendering HTML result.
+     *
+     * @throws ViewNotFoundException If the view file does not exist.
+     */
+    public function renderHtml(string $view, array $viewParameters = [], array $layoutParameters = []): string
+    {
+        $content = $this->view
+            ->withContextPath($this->template->viewPath)
+            ->render($view, $viewParameters);
+
+        if ($this->template->htmlLayout === null) {
+            return $content;
+        }
+
+        $layoutParameters['content'] = $content;
+        return $this->view
+            ->withContextPath($this->template->viewPath)
+            ->render($this->template->htmlLayout, $layoutParameters);
+    }
+
+    /**
+     * Renders the text view specified with optional parameters and layout.
+     *
+     * @param string $view The view name of the view file.
+     * @param array $viewParameters The parameters (name-value pairs) that will be extracted and available in
+     * the view file.
+     * @param array $layoutParameters The parameters (name-value pairs) that will be extracted and available in
+     * the layout file.
+     *
+     * @see View::render()
+     *
+     * @return string The rendering text result.
+     *
+     * @throws ViewNotFoundException If the view file does not exist.
+     */
+    public function renderText(string $view, array $viewParameters = [], array $layoutParameters = []): string
+    {
+        $content = $this->view
+            ->withContextPath($this->template->viewPath)
+            ->render($view, $viewParameters);
+
+        if ($this->template->textLayout === null) {
+            return $content;
+        }
+
+        $layoutParameters['content'] = $content;
+        return $this->view
+            ->withContextPath($this->template->viewPath)
+            ->render($this->template->textLayout, $layoutParameters);
+    }
+
+    /**
+     * Generates a text body from an HTML body.
+     *
+     * @param string $html The HTML body.
+     *
+     * @return string The text body.
+     */
+    private function generateTextBodyFromHtml(string $html): string
+    {
+        if (preg_match('~<body[^>]*>(.*?)</body>~is', $html, $match)) {
+            $html = $match[1];
+        }
+
+        // remove style and script
+        $html = preg_replace('~<((style|script))[^>]*>(.*?)</\1>~is', '', $html);
+
+        // strip all HTML tags and decode HTML entities
+        $text = html_entity_decode(strip_tags($html), ENT_QUOTES | ENT_HTML5);
+
+        // improve whitespace
+        $text = preg_replace("~^[ \t]+~m", '', trim($text));
```

Comment:
Does it make sense to remove spaces from the end of lines?

```suggestion
        $text = preg_replace("~^[ \t]+|[ \t]+$~m", '', trim($text));
```